### DISCUSSION TITLE
Supabase OAuth not working with the Next.js App Router

### DISCUSSION THREAD

**Discussion by clementgateaud** [supabase/supabase#15339] [Questions] [answered]
Hello 👋

I created a NextJS 13 (with "app" directory) and set up [supabase auth through cookies](https://supabase.com/docs/guides/auth/auth-helpers/nextjs), it's working well with email/password.
However, I don't manage to make OAuth work with providers (I tried Google and GitHub).

I have a "Sign in with GitHub" button triggering this function:

``` 
const handleSignInWithGithub = async () => {
  await supabase.auth.signInWithOAuth({
    provider: "github",
    options: {
      redirectTo: `${location.origin}/auth/callback`,
    },
  });
};
```
I followed the docs to set up everything on the GitHub side (add supabse callback URL, and add to supabase github client id and secret).

When clicking on my "Sign in with GitHub" button, I can fill in my GitHub credentials, and then get redirected back to my app, but it doesn't create a session, and I don't see a new user on my Supabase users table. 
However, it creates a cookie `sb-<My_Supabase_Url>-token-code-verifier`. But that's it.

As explained in the docs, I create a route `src/app/auth/callback/route.ts` that's supposed to handle code exchange for sessio:

```
import { Database } from "@/database.types";
import { createRouteHandlerClient } from "@supabase/auth-helpers-nextjs";
import { cookies } from "next/headers";
import { NextResponse } from "next/server";

export async function GET(request: Request) {
  const requestUrl = new URL(request.url);
  const code = requestUrl.searchParams.get("code");

  if (code) {
    const supabase = createRouteHandlerClient<Database>({ cookies });
    await supabase.auth.exchangeCodeForSession(code);
  }

  // URL to redirect to after sign in process completes
  return NextResponse.redirect(requestUrl.origin);
}
```

Am I missing something or doing anything wrong ?

If anyone had the same issue and/or know what I could have done wrong, it would be amazing 🙏

**Comment by lookevink** [comment]
I pulled hair all day over some other issue with auth/callback route that I will make a pr for. the tutorial doesn't handle all of edge cases especially if something's going on with your environment.

when you have a token-code-verifier but no auth-token, it means that the code exchange did not happen successfully. basically the helper takes the verifier token from cookies as well as the code from url to exchange for a session token that it then stores in cookies again.  Hard to say exactly what's causing your issue but here are 2 things I'd ensure are properly configured. 
1. In your SB & GH portals, your Redirect URLs include hostname that you're using. it's most likely https://localhost:3000 for local dev
2. you're not switching browsers during the auth process. after SB completes the authentication, it redirects you to hostname/auth/callback. the redirect causes a GET request to the endpoint, which accesses the aforementioned code and verifier to exchange for a session. 

maybe post some more information from console & network tabs so I can better help?

**Comment by clementgateaud** [comment, marked as answer]
Thanks a lot @lookevink  for taking time to answer and give me some explanations, I finally found out what the issue was !

The issue was coming from my `middleware.ts` file 🤦‍♂️
I had added code in my middleware to make sure the url was containing the user's locale ("/en", "/fr" etc.) and I was redirecting the user if the locale was not specified.

So when trying to access `/auth/callback`, the middleware was redirecting to another url 😅
I added `/auth` urls as exceptions for these URL transformations, and now everything is working fine !
